driven/generic/adapter.py

```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
+        self.changes = {
+            'added': {'reactions': set(), 'metabolites': set()},  # reaction contain information about genes
+            'removed': {'genes': set()},
+        }
+        self.apply_changes(genotype_changes)
+
+    def apply_changes(self, genotype_changes):
+        """Apply genotype changes on initial model
+
+        :param genotype_changes: gnomic.Genotype
+        :return:
+        """
+        for change in genotype_changes.changes():
+            if isinstance(change, gnomic.Mutation):
+                self.apply_mutation(change)
+            if isinstance(change, gnomic.Plasmid):
+                self.add_plasmid(change)
+
+    def apply_mutation(self, mutation):
+        """Apply mutations on initial model
+
+        :param mutation: gnomic.Mutation
+        :return:
+        """
+        if mutation.old:
+            for feature in mutation.old.features():
+                self.knockout_gene(feature)
+        if mutation.new:
+            for feature in mutation.new.features():
+                self.add_gene(feature)
+
+    def add_plasmid(self, plasmid):
+        """Add plasmid features to the initial model.
+        No plasmid instance in cameo, so changes are made in model genes and reactions directly
+
+        :param plasmid: gnomic.Plasmid
+        :return:
+        """
+        for feature in plasmid.features():
+            self.add_gene(feature)
+
+    def knockout_gene(self, feature):
+        """Perform gene knockout.
+        Use feature name as gene name
+
+        :param feature: gnomic.Feature
+        :return:
+        """
+        gene = self.model.genes.query(feature.name, attribute="name")
+        if gene:
+            gene[0].knock_out()
+            self.changes['removed']['genes'].add(gene[0])
+            logger.info('Gene knockout: {}'.format(gene[0].name))
+        else:
+            logger.info('Gene for knockout is not found: {}'.format(feature.name))
+
+    def add_gene(self, feature):
+        """Perform gene insertion.
+        Find all the reactions associated with this gene using KEGGClient and add them to the model
+
+        :param feature: gnomic.Feature
+        :return:
+        """
+        logger.info('Add gene: {}'.format(feature.name))
+        identifier = feature_id(feature)
+        if self.model.genes.query(identifier, attribute='name'):  # do not add if gene is already there
+            logger.info('Gene {} exists in the model'.format(feature.name))
+            return
+        for reaction_id, equation in self.genes_to_reactions.get(identifier, {}).items():
+            self.add_reaction(reaction_id, equation, identifier)
+        logger.info('Gene added: {}'.format(identifier))
+
+    def add_reaction(self, reaction_id, equation, gene_name):
+        """Add new reaction by rn ID from equation, where metabolites defined by kegg ids.
+
+        :param reaction_id: reaction rn ID
+        :param equation: equation string, where metabolites are defined by kegg ids
```

Comment:
Are KEGG ids the only valid ids? If yes, you can make a check before hand. KEGG ids match the expression /C\d{5}/